### PR TITLE
gitserver: Implement GetCommit as RPC method

### PR DESCRIPTION
**Pull request by eseliger**
Analogous to the other migration PRs, this PR migrates the GetCommit method to a separate gRPC endpoint.

## Test plan

Manually played around with instance and added tests.

### CODE REVIEW

File: cmd/gitserver/internal/git/iface.go [hunk 1]
```diff
@@ -73,3 +78,11 @@ type BlameHunkReader interface {
...
 	Close() error
 }
+
+// GitCommitWithFiles wraps a gitdomain.Commit and adds a list of modified files.
+// Modified files are only populated when requested.
+// This data is required for sub repo permission filtering.
+type GitCommitWithFiles struct {
+	*gitdomain.Commit
+	ModifiedFiles []string
+}
```
Comment (eseliger) [inline cmd/gitserver/internal/git/iface.go:88]: I don't love this, but subrepo permissions somewhat force us to do this right now. Maybe we can revisit the implementation of SRP after all the migrations are done, and see if this is actually necessary.

File: internal/gitserver/gitdomain/common_test.go [hunk 2]
```diff
@@ -6,6 +6,43 @@ import (
 	"github.com/stretchr/testify/assert"
 )
 
+func TestMessage(t *testing.T) {
```
Comment (eseliger) [inline internal/gitserver/gitdomain/common_test.go:9]: Move here, felt wrong in the client package where the type is not defined..

File: internal/gitserver/gitdomain/common_test.go [hunk 3]
```diff
@@ -110,19 +154,75 @@ func TestIsAbsoluteRevision(t *testing.T) {
...
+		}
+
+		return true
+	}, nil)
+	if err != nil {
+		t.Fatalf("unexpected diff (-want +got):\n%s", diff)
+	}
+}
+
+func TestRoundTripCommit(t *testing.T) {
```
Comment (eseliger) [inline internal/gitserver/gitdomain/common_test.go:187]: yay, thanks Geoffrey for teaching me how to properly test this although it contains a time.Time :)

File: cmd/gitserver/internal/git/gitcli/odb.go [hunk 4]
```diff
@@ -5,12 +5,139 @@ import (
...
+		if errors.As(err, &e) && e.ExitStatus == 128 && bytes.Contains(e.Stderr, []byte("fatal: bad object")) {
+			return nil, &gitdomain.RevisionNotFoundError{Repo: g.repoName, Spec: string(commit)}
+		}
+
+		return nil, err
+	}
+
+	c, err := parseCommitLogOutput(bytes.TrimPrefix(rawCommit, []byte{'\x1e'}))
+	if err != nil {
+		return nil, err
```
Comment (ggilmore) [inline cmd/gitserver/internal/git/gitcli/odb.go:52]: maybe wrap the error here?

File: cmd/gitserver/internal/server_grpc.go [hunk 5]
```diff
@@ -848,6 +848,84 @@ func (gs *grpcServer) MergeBase(ctx context.Context, req *proto.MergeBaseRequest
...
+		})
+		if err != nil {
+			return nil, err
+		}
+		return nil, s.Err()
+	}
+
+	subRepoPermsEnabled, err := authz.SubRepoEnabledForRepo(ctx, gs.subRepoChecker, repoName)
+	if err != nil {
+		return nil, err
```
Comment (ggilmore) [inline cmd/gitserver/internal/server_grpc.go:885]: wrap the error here?

File: cmd/gitserver/internal/server_grpc.go [hunk 6]
```diff
@@ -848,6 +848,84 @@ func (gs *grpcServer) MergeBase(ctx context.Context, req *proto.MergeBaseRequest
...
+	commit, err := backend.GetCommit(ctx, api.CommitID(req.GetCommit()), subRepoPermsEnabled)
+	if err != nil {
+		var e *gitdomain.RevisionNotFoundError
+		if errors.As(err, &e) {
+			s, err := status.New(codes.NotFound, "revision not found").WithDetails(&proto.RevisionNotFoundPayload{
+				Repo: req.GetRepoName(),
+				Spec: e.Spec,
+			})
+			if err != nil {
+				return nil, err
```
Comment (ggilmore) [inline cmd/gitserver/internal/server_grpc.go:899]: wrap this error?

File: cmd/gitserver/internal/server_grpc.go [hunk 7]
```diff
@@ -848,6 +848,84 @@ func (gs *grpcServer) MergeBase(ctx context.Context, req *proto.MergeBaseRequest
...
+			}
+			return nil, s.Err()
+		}
+		gs.svc.LogIfCorrupt(ctx, repoName, err)
+		// TODO: Better error checking.
+		return nil, err
```
Comment (ggilmore) [inline cmd/gitserver/internal/server_grpc.go:905]: wrap this error here?

File: cmd/gitserver/internal/server_grpc.go [hunk 8]
```diff
@@ -848,6 +848,84 @@ func (gs *grpcServer) MergeBase(ctx context.Context, req *proto.MergeBaseRequest
...
+	}
+
+	hasAccess, err := hasAccessToCommit(ctx, repoName, commit.ModifiedFiles, gs.subRepoChecker)
+	if err != nil {
+		return nil, err
```
Comment (ggilmore) [inline cmd/gitserver/internal/server_grpc.go:910]: wrap this error here?

File: cmd/gitserver/internal/server_grpc.go [hunk 9]
```diff
@@ -848,6 +848,84 @@ func (gs *grpcServer) MergeBase(ctx context.Context, req *proto.MergeBaseRequest
...
+	}
+
+	if !hasAccess {
+		s, err := status.New(codes.NotFound, "revision not found").WithDetails(&proto.RevisionNotFoundPayload{
+			Repo: req.GetRepoName(),
+			Spec: req.GetCommit(),
+		})
+		if err != nil {
+			return nil, err
```
Comment (ggilmore) [inline cmd/gitserver/internal/server_grpc.go:919]: wrap this error here?

**Review by ggilmore — APPROVED** [review]
just some error wrapping questions

**Pull request merged**